File: Lista de Exercicio Paradigmas/Lista Java/Ex8.java
import java.util.ArrayList;
import java.util.List;

class Professor {
    private String nome;
    private List<Escola> escolas;

    public Professor(String nome) {
        this.nome = nome;
        this.escolas = new ArrayList<>();
    }

    public void adicionarEscola(Escola escola) {
        if (!escolas.contains(escola)) {
            escolas.add(escola);
            escola.adicionarProfessor(this);
        }
    }

    public List<String> listarEscolas() {
        List<String> nomesEscolas = new ArrayList<>();
        for (Escola escola : escolas) {
            nomesEscolas.add(escola.getNome());
        }
        return nomesEscolas;
    }

    public String getNome() {
        return nome;
    }
}

class Escola {
    private String nome;
    private List<Professor> professores;

    public Escola(String nome) {
        this.nome = nome;
        this.professores = new ArrayList<>();
    }

    public void adicionarProfessor(Professor professor) {
        if (!professores.contains(professor)) {
            professores.add(professor);
        }
    }

    public List<String> listarProfessores() {
        List<String> nomesProfessores = new ArrayList<>();
        for (Professor professor : professores) {
            nomesProfessores.add(professor.getNome());
        }
        return nomesProfessores;
    }

    public String getNome() {
        return nome;
    }
}

public class Main {
    public static void main(String[] args) {
        Escola escola1 = new Escola("Escola Primária");
        Escola escola2 = new Escola("Escola Secundária");

        Professor professor1 = new Professor("Carlos Silva");
        Professor professor2 = new Professor("Maria Oliveira");

        professor1.adicionarEscola(escola1);
        professor1.adicionarEscola(escola2);
        professor2.adicionarEscola(escola1);

        System.out.println("Professores da " + escola1.getNome() + ": " + escola1.listarProfessores());
        System.out.println("Escolas que o " + professor1.getNome() + " leciona: " + professor1.listarEscolas());
    }
}

File: Lista de Exercicio Paradigmas/Lista Python/Ex8.py
class Professor:
    def __init__(self, nome):
        self.nome = nome
        self.escolas = []  

    def adicionar_escola(self, escola):
        if escola not in self.escolas:
            self.escolas.append(escola)
            escola.adicionar_professor(self)  

    def listar_escolas(self):
        return [escola.nome for escola in self.escolas]

class Escola:
    def __init__(self, nome):
        self.nome = nome
        self.professores = []  

    def adicionar_professor(self, professor):
        if professor not in self.professores:
            self.professores.append(professor)

    def listar_professores(self):
        return [professor.nome for professor in self.professores]

escola1 = Escola('Escola Primária')
escola2 = Escola('Escola Secundária')

professor1 = Professor('Carlos Silva')
professor2 = Professor('Maria Oliveira')

professor1.adicionar_escola(escola1)
professor1.adicionar_escola(escola2)
professor2.adicionar_escola(escola1)

print(f'Professores da {escola1.nome}: {escola1.listar_professores()}')
print(f'Escolas que o {professor1.nome} leciona: {professor1.listar_escolas()}')